src/state/categories/selectors.js

```diff
@@ -0,0 +1,109 @@
+import { createSelector } from 'reselect';
+import { get } from 'lodash';
+import { getLocale } from 'state/locale/selectors';
+
+export const getCategories = state => state.categories;
+export const getCategoriesIdList = state => state.categories.list;
+export const getCategoriesMap = state => state.categories.map;
+export const getChildrenMap = state => state.categories.childrenMap;
+export const getStatusMap = state => state.categories.statusMap;
+export const getTitlesMap = state => state.categories.titlesMap;
+export const getSelected = state => get(state.categories, 'selected', {});
+export const getSelectedRefs = state => get(state.categories, 'selected.references', {});
+export const getReferenceKeyList = state => get(state.categories, 'selected.referenceKeyList', []);
+export const getReferenceMap = state => get(state.categories, 'selected.referenceMap', {});
+
+const CATEGORY_STATUS_DEFAULTS = {
+  expanded: false,
+  loading: false,
+  loaded: false,
+};
+
+const getCategoriesOrder = (categoriesChildren) => {
+  const fifo = ['home'];
+  const sorted = [];
+  while (fifo.length) {
+    const curCategoryCode = fifo.pop();
+    sorted.push(curCategoryCode);
+    if (categoriesChildren[curCategoryCode]) {
+      categoriesChildren[curCategoryCode]
+        .slice()
+        .reverse()
+        .forEach((code) => {
+          fifo.push(code);
+        });
+    }
+  }
+  return sorted;
+};
+
+const isVisible = (categoryCode, categories, categoriesStatus) => {
+  let curCategoryCode = categoryCode;
+  if (categories[curCategoryCode]) {
+    while (curCategoryCode !== 'home') {
+      curCategoryCode = categories[curCategoryCode].parentCode;
+      if (categoriesStatus[curCategoryCode] && !categoriesStatus[curCategoryCode].expanded) {
+        return false;
+      }
+    }
+    return true;
+  }
+  return false;
+};
+
+const getDepth = (categories, categoryCode) => {
+  let curCategoryCode = categoryCode;
+  let depth = 0;
+  if (categories[curCategoryCode]) {
+    while (curCategoryCode !== 'home') {
```

Comment:
why a while?

Reply:
A variable *curCategoryCode* is declared on line 55. But, it is then re-assigned on line 59 inside *while* scope, here: `curCategoryCode = categories[curCategoryCode].parentCode;` . So, it takes the child and crawls up to parent, then it's parent's parent etc up until *home* is not reached. The same concept is in *isVisible* function.